### ISSUE TITLE
question about the implementation

### ISSUE THREAD

**Issue by qianyizhang** [ZiweiWangTHU/ISL#1]
Hi,
I am a bit confused about the implementation:
1. this is obviously wrong. Perhaps it doesn't hurt as much to dataset with large labels, but i believe it would be detrimental to SVHN and CIFAR10.
https://github.com/ZiweiWangTHU/ISL/blob/main/lib/LinearAverage.py#L156-L157

2. whats the point of aliasing sampling when you have a uniform distribution throughout the experiment?
https://github.com/ZiweiWangTHU/ISL/blob/main/lib/LinearAverage.py#L55-L56

3. you make a somewhat confusing comment, i believe the size of output is [batch_size, data_num]
https://github.com/ZiweiWangTHU/ISL/blob/main/imagenet_main.py#L219

**Comment by Wuziyi616** [comment]
Thank you for your interest in our work, and your effort in pointing out the problems.

1. Yes that's definitely wrong. We are looking into it and see if our experiment code also has this error. Because the small datasets experiments were performed by me (about March, 2020) and ImageNet experiments were performed by the second author later (about Sept., 2020), I need to take some time to find the code.
2. That's simply because I copy & paste the code [here](https://github.com/zhirongw/lemniscate.pytorch/blob/f7cfe298357cb2b169cd59eb540aca24bed1f9b8/lib/NCEAverage.py#L78). But later we found it useless and we just use uniform sampling. Indeed we should add a reference to their code in the README
3. If I remember correctly, this is because `num_data == num_neg+num_pos`, and `num_pos == 1`. So I write `num_neg+1` here. That's a bit confusing. We will correct it later

**Comment by qianyizhang** [comment]
> Thank you for your interest in our work, and your effort in pointing out the problems.
> 
> 1. Yes that's definitely wrong. We are looking into it and see if our experiment code also has this error. Because the small datasets experiments were performed by me (about March, 2020) and ImageNet experiments were performed by the second author later (about Sept., 2020), I need to take some time to find the code.
> 2. That's simply because I copy & paste the code [here](https://github.com/zhirongw/lemniscate.pytorch/blob/f7cfe298357cb2b169cd59eb540aca24bed1f9b8/lib/NCEAverage.py#L78). But later we found it useless and we just use uniform sampling. Indeed we should add a reference to their code in the README
> 3. If I remember correctly, this is because `num_data == num_neg+num_pos`, and `num_pos == 1`. So I write `num_neg+1` here. That's a bit confusing. We will correct it later

on #3, no. If you trace the function calls, the `lemniscate ` is actually: 
https://github.com/ZiweiWangTHU/ISL/blob/aa1de9f8f3ceb8db69fb1f09c88006bd8856d634/imagenet_main.py#L148

https://github.com/ZiweiWangTHU/ISL/blob/aa1de9f8f3ceb8db69fb1f09c88006bd8856d634/lib/LinearAverage.py#L70

https://github.com/ZiweiWangTHU/ISL/blob/aa1de9f8f3ceb8db69fb1f09c88006bd8856d634/lib/LinearAverage.py#L16

which is the inner product between [batch * emb, data_num * emb], where data_num is the size of entire dataset.
In another word, I think the memory bank is more 'online' than 'offline', since it's been utilized in full and partially updated in each forward pass.
As a result, it takes 2gb + for imagenet, and makes this method inapplicable to even larger dataset.

**Comment by Wuziyi616** [comment]
Oh, yes, you are right. Sorry I haven't been doing this for a long time (about 1 year). If I remember correctly, we use full dataset following AND and PAD. I have considered using the InfoNCE form as InstDist, but I didn't find a good approximation for calculating the mined neighbor terms.

**Comment by qianyizhang** [comment]
ok then. 
please fix #1 and see what happens. It's a sound idea, but sucks it's being right (as for the improvements) for the wrong reason...

**Comment by Wuziyi616** [comment]
I am not in that lab any more and I will ask my co-workers if they can spare some time to run the code. Thanks for pointing out the error any way!

**Comment by Wuziyi616** [comment]
I've retrained the model, the performance on small datasets like CIFAR10 doesn't change much (<0.2% difference for ResNet18). On ImageNet, the performance consistently improves for 0.2~0.5% (ResNet18). I believe if we tune the hyper-parameters more carefully, we can expect more improvement.

**Comment by qianyizhang** [comment]
> I've retrained the model, the performance on small datasets like CIFAR10 doesn't change much (<0.2% difference for ResNet18). On ImageNet, the performance consistently improves for 0.2~0.5% (ResNet18). I believe if we tune the hyper-parameters more carefully, we can expect more improvement.

thanks for the update. 
But honestly speaking, does this phenomena match your expectation? I mean you fixed an error that potentially affects small dataset with greater impact, but somehow it degrades it to a miner extent and improves the larger one...
it makes me to wonder if you are being right for the wrong reason (aka, performance boost coming from elsewhere than whats claimed)

**Comment by Wuziyi616** [comment]
Oh sorry, I think what I was saying is a bit confusing. On CIFAR fixing the error also improves the performance (`<0.2 difference` means increases like `0.15%` something, I should have used the word `improvement` here). But since the dataset is small and performance is already very high, the gain is just marginal. On the other hand, ImageNet is large and we get large improvements. Our baseline is directly taken from the Instance Discrimination paper's open source code.
